### PR TITLE
Documentation: Overview of SSA

### PR DESCRIPTION
**Pull request by Degiorgio**
Specifically, this commit modifies goto-symex/README.md by elaborating on the roles played by symex_target_equationt and symex_targett.

Note':  README.md    is missing a description and example  of level-0  SSA versioning (to come later). 
Note'': The code-level documentation of  the symex_target.{h.cpp} and symex_target_equation.{h,cpp} is still in progress.

### CODE REVIEW

File: src/goto-symex/README.md [hunk 1]
```diff
@@ -15,12 +15,22 @@ when analysis is incomplete.  The symbolic execution includes constant
...
 
-The output of the symbolic execution is a system of equations; an object
-containing a list of `symex_target_elements`, each of which are
-equalities between `expr` expressions. See `symex_target_equation.h`.
-The output is in static, single assignment (SSA) form, which is *not*
-the case for goto-programs.
+The output of symbolic execution is a system of equations; an object
+of type `symex_target_equationt`, containing a list of
+`symex_target_equationt::SSA_stept`, each of which are equalities
+between `exprt` expressions. This list is constructed incrementally as
```
Comment (tautschnig) [inline src/goto-symex/README.md]: Only assignments translate to equations, other instructions need not yield those (consider assumptions, assertions).

File: src/goto-symex/README.md [hunk 2]
```diff
@@ -95,47 +105,73 @@ execution run does not add any paths to the workqueue but rather merges
...
+1. Every variable is *assigned exactly once*.
+2. Every variable must be *defined* before use.
+
+In-order to convert a goto-program to SSA form all variables are
+versioned (renamed) through the addition of a _suffix_.
+
+There are three “levels” of versioning:
+
+**Level 2 (L2):** variables are versioned every time they are encountered
+in a function.
```
Comment (tautschnig) [inline src/goto-symex/README.md]: This is the SSA index introduced in the 1988 SSA paper.

File: src/goto-symex/README.md [hunk 3]
```diff
@@ -95,47 +105,73 @@ execution run does not add any paths to the workqueue but rather merges
...
+
+**Level 1 (L1):** variables are versioned every time the functions that
+contain those variables are called.
```
Comment (tautschnig) [inline src/goto-symex/README.md]: I'm not sure there is standard terminology for this, but the code refers to this index as a frame.

File: src/goto-symex/README.md [hunk 4]
```diff
@@ -95,47 +105,73 @@ execution run does not add any paths to the workqueue but rather merges
...
+
+**Level 0 (L0):** variables are versioned every time a new thread
+containing those variables are spawned.
```
Comment (tautschnig) [inline src/goto-symex/README.md]: This is based on "Concurrent static single assignment form and constant propagation for explicitly parallel programs" by Lee, Midkiff, Padua. Might call it "thread index."

File: src/goto-symex/README.md [hunk 5]
```diff
@@ -150,6 +186,10 @@ incremented (from 1 to 2) and the L0 counter is reset (back to 2, after
 having been incremented up to 4). The L0 counter then increases every
```
Comment (tautschnig) [inline src/goto-symex/README.md]: The use of "L0" seems wrong here.

File: src/goto-symex/README.md [hunk 6]
```diff
@@ -95,47 +105,73 @@ execution run does not add any paths to the workqueue but rather merges
...
-containing those variables are spawned. We can inspect the renamed
-variable names with the –show-vcc flag, which prints a string with the
-following format: `%%s!%%d@%%d#%%d`. The string field is the variable name,
-and the numbers after the !, @, and # are the L0, L1, and L2 suffixes
-respectively. The following examples illustrate Level 1 and 2 renaming:
+---
+\section static-single-assignment Static Single Assignment (SSA) Form
+
+**Key classes:**
+* symex_targett
```
Comment (peterschrammel) [inline src/goto-symex/README.md]: link

File: src/goto-symex/README.md [hunk 7]
```diff
@@ -95,47 +105,73 @@ execution run does not add any paths to the workqueue but rather merges
...
+**Level 2 (L2):** variables are versioned every time they are encountered
+in a function.
+
+**Level 1 (L1):** variables are versioned every time the functions that
+contain those variables are called.
+
+**Level 0 (L0):** variables are versioned every time a new thread
+containing those variables are spawned.
+
+We can inspect the versioned variable names with the **–show-vcc** flag,
```
Comment (peterschrammel) [inline src/goto-symex/README.md]: or program-only

File: src/goto-symex/README.md [hunk 8]
```diff
@@ -150,6 +186,10 @@ incremented (from 1 to 2) and the L0 counter is reset (back to 2, after
 having been incremented up to 4). The L0 counter then increases every
 time we access x as we walk through the function.
 
+\subsection L0 Level 0 versioning
+
+TODO
```
Comment (peterschrammel) [inline src/goto-symex/README.md]: This is for threads.

File: src/goto-symex/README.md (hunk 1 above)
Comment (Degiorgio) [reply src/goto-symex/README.md]: Will clarify this.

File: src/goto-symex/README.md (hunk 2 above)
Comment (Degiorgio) [reply src/goto-symex/README.md]: Will add a citation.

File: src/goto-symex/README.md (hunk 4 above)
Comment (Degiorgio) [reply src/goto-symex/README.md]: See earlier comment,  Will also add  a citation.

File: src/goto-symex/README.md (hunk 8 above)
Comment (Degiorgio) [reply src/goto-symex/README.md]: yes this only  relevant   when threads are present, will add comment.

File: src/goto-symex/README.md (hunk 3 above)
Comment (Degiorgio) [reply src/goto-symex/README.md]: @tautschnig  the previous, version,  was using the terminology, 'renamed' , I thought 'versioned' is more appropriate, however 'indexed' is also fine.   Do you have a preference?  (assuming as you say there isn't a prevailing terminology adopted by primary literature) .

File: src/goto-symex/README.md (hunk 5 above)
Comment (Degiorgio) [reply src/goto-symex/README.md]: yep fixing.

**Comment by Degiorgio** [comment]
@peterschrammel  @tautschnig  latest commit should address the issues you raised.

File: src/goto-symex/README.md (hunk 3 above)
Comment (tautschnig) [reply src/goto-symex/README.md]: How about rewording it as follows: "frame index incremented every time a function containing the variable is called?"

**Review by tautschnig — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]
